AMDGPU: fix isSafeToSink expecting exactly one predecessor

petar-avramovic opened the pull request:
isSafeToSink needs to check if machine cycle has divergent exit branch
but first it needs the MBB that contains cycle exit branch.
Early-tailduplication can delete exit block created by structurize-cfg
so there is still exactly one cycle exit block but the new cycle exit
block can have multiple predecessors.
Simplify search for MBBs that contain cycle exit branch by introducing
helper method getExitingBlocks in GenericCycle.

Fixes #89200

jayfoad commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -217,10 +217,20 @@ bool SIInstrInfo::isSafeToSink(MachineInstr &MI,
         assert(ExitBlocks.size() == 1);
-        assert(ExitBlocks[0]->getSinglePredecessor());
-
+        // After structurize-cfg, cycle exit block should have exactly one
```
Isn't it simpler to write this code so that it makes no assumptions about structurized CFGs? E.g.
```c++
  FromCycle->getExitBlocks(ExitBlocks);
  for (ExitBlock : ExitBlocks)
    for (Pred : ExitBlock->predecessors())
      if (FromCycle->contains(Pred) && hasDivergentBranch(Pred))
        return false;
```

petar-avramovic replied:
OK, write it like that for simplicity and delete asserts. Can we leave some of the comments? I don't like fully hiding expected CFG as it could give wrong impression that we can deal with cycle with two divergent exits or similar

jayfoad replied:
> it could give wrong impression that we can deal with cycle with two divergent exits or similar

*This* code can deal with all those cases, can't it?

petar-avramovic replied:
Your proposal without asserts would return something even if input was not "structurized-cfg". Do we want to return something even when CFG is something we don't handle?

jayfoad replied:
Yes, if *this* code can easily handle arbitrary CFGs then it should do so.

If other parts of the code cannot handle arbitrary CFGs, then they should detect the problems.

ssahasra replied:
Yes, whether we return a block should be independent of structrize-cfg. But we should keep the comment for our own future help.

ssahasra reviewed (approved):
LGTM, but please wait for @jayfoad

ruiling commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -213,17 +213,20 @@ bool SIInstrInfo::isSafeToSink(MachineInstr &MI,
+        // After structurize-cfg, there should be exactly one cycle exit. Also,
+        // cycle exit block should have exactly one predecessor, the cycle exit.
+        // Early-tailduplication can removed that block so we have to search for
+        // predecessor that is in cycle.
```
The comment here is not useful anymore. The logic should work in general CFG. Meanwhile, I would like to see the code cleaner. Instead of getting exit blocks and check its predecessors. Let's just check against cycle exiting blocks.
```
FromCycle->getExitingBlocks(ExitingBlocks);
for (auto *Exiting : ExistingBlocks)
  if (hasDivergentBranch(Exiting))
    return false;
```

jayfoad replied:
`LoopInfo` has `getExitingBlocks` but `CycleInfo` does not.

I would be happy with removing the comment but @ssahasra suggested to keep it.

petar-avramovic replied:
I like getExitingBlocks more.
btw, is that the formal name for blocks that contain branch that exits the cycle?

petar-avramovic replied:
Should I switch to using LoopInfo then? For our targets cycles at this point should all be natural loops.

arsenm replied:
I maintain it is fully invalid to expect anything out of the MIR CFG. If it isn't enforced by the machine verifier, it isn't required. You cannot assume natural loops

petar-avramovic committed:
AMDGPU: fix isSafeToSink expecting exactly one predecessor #89224

isSafeToSink needs to check if machine cycle has divergent exit branch
but first it needs the MBB that contains cycle exit branch.
Early-tailduplication can delete exit block created by structurize-cfg
so there is still exactly one cycle exit block but the new cycle exit
block can have multiple predecessors.
Simplify search for MBBs that contain cycle exit branch by introducing
helper method getExitingBlocks in GenericCycle.

jayfoad commented on llvm/include/llvm/ADT/GenericCycleImpl.h:
```diff
@@ -66,6 +66,19 @@ void GenericCycle<ContextT>::getExitBlocks(
+  for (BlockT *Block : blocks()) {
+    for (BlockT *Succ : successors(Block)) {
+      if (!contains(Succ))
+        TmpStorage.push_back(Block);
```
`break` from the inner loop here.

jayfoad commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -214,13 +214,12 @@ bool SIInstrInfo::isSafeToSink(MachineInstr &MI,
 
         // FromCycle has divergent exit condition.
-        if (hasDivergentBranch(ExitBlocks[0]->getSinglePredecessor())) {
+        if (hasDivergentBranch(ExitingBlocks[0])) {
```
Please return false if _any_ exiting block has a divergent branch

jayfoad commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -214,14 +214,14 @@ bool SIInstrInfo::isSafeToSink(MachineInstr &MI,
+        for (MachineBasicBlock *ExitingBlock : ExitingBlocks) {
+          if (hasDivergentBranch(ExitingBlock)) {
+            return false;
+          }
```
Nit: don't need braces around a single physical line.

jayfoad reviewed (commented):
Looks reasonable to me but please check with @ssahasra and @ruiling.

ruiling reviewed (approved):
Thanks for fixing this. LGTM

ruiling commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -214,14 +214,13 @@ bool SIInstrInfo::isSafeToSink(MachineInstr &MI,
       // does not contain SuccToSinkTo and also has divergent exit condition.
       while (FromCycle && !FromCycle->contains(ToCycle)) {
         // After structurize-cfg, there should be exactly one cycle exit.
```
I just notice the old comment is still there, please help remove it. Thanks!
